onmt/utils/optimizers.py

```diff
@@ -2,87 +2,108 @@
 import torch
 import torch.optim as optim
 from torch.nn.utils import clip_grad_norm_
-from onmt.utils import use_gpu
 import operator
 import functools
 from copy import copy
 from math import sqrt
 
 
-def build_optim(model, opt, checkpoint):
-    """ Build optimizer """
-    saved_optimizer_state_dict = None
-
-    if opt.train_from and opt.reset_optim != 'all':
-        optim = checkpoint['optim']
-        # We need to save a copy of optim.optimizer.state_dict() for setting
-        # the, optimizer state later on in Stage 2 in this method, since
-        # the method optim.set_parameters(model) will overwrite
-        # optim.optimizer, and with ith the values stored in
-        # optim.optimizer.state_dict()
-        if opt.reset_optim != 'states':
-            saved_optimizer_state_dict = optim.optimizer.state_dict()
-            if opt.reset_optim == 'keep_states':
-                optim.method = opt.optim
-                optim.learning_rate = opt.learning_rate
-                optim.original_lr = opt.learning_rate
-                optim.max_grad_norm = opt.max_grad_norm
-                optim.lr_decay = opt.learning_rate_decay
-                optim.start_decay_steps = opt.start_decay_steps
-                optim.decay_steps = opt.decay_steps
-                optim.betas = [opt.adam_beta1, opt.adam_beta2]
-                optim.adagrad_accum = opt.adagrad_accumulator_init
-                optim.decay_method = opt.decay_method
-                optim.warmup_steps = opt.warmup_steps
-                optim.model_size = opt.rnn_size
+def make_optimizer(model, opt):
```

Comment:
Why the change to the function name? I am open to all sorts of changes with how objects are created, but all the other ones start with 'build' and this breaks that consistency.

Reply:
This is a new function that creates a `torch.optim.Optimizer` instance. The previous `build_optim` function has been moved to `Optimizer.from_options` to avoid confusion.

Maybe we should rename the `Optimizer` class to something else to avoid confusion with the PyTorch class with the same name?

Reply:
I agree the new structure make things unclear.
IMO:
let's keep build_optim to build the pytorch Optimizer
let's rename our class Optimizer to Optim_wrapper
I don't  like the from_options name. Would the call Optim_wrapper.build() make sense ?
open to other ideas.

Reply:
I don't much like the `OptimWrapper` name because it's presenting things at a level of abstraction that is not useful to people who work on machine translation. Optimizers I understand. Wrappers, managers, builders -- things like this are not so interesting to me and make my life harder.

I like the name `from_options` (or maybe `from_opt` if it's clearer). It is a pattern for constructor naming that is well established elsewhere (numpy arrays have `fromfile`, `fromfunction`, `fromiter`, etc.; pandas dataframes have `from_dict`, `from_items`, etc.). I think this is much better than a name like `build()`, which is completely uninformative.

Reply:
I am not fighting this. But if so, there are plenty of places where we would need to rename to make things consistent all the way through. So far, we used build to create instances.
As for the Optimizer there is definitely an overlap with the Pytorch native optimizer class. Zero abstraction in names when there is, is not good either.

Reply:
That won't happen soon if we keep 1) Adafactor 2) SparseAdam (which is not the same as the legacy pytorch SparseAdam Optimizer that supports onyl sparse tensors). We could drop these 2, but then it's another story. In the meantime, I'm fine too leave things as they are, not optimal but ok.

Reply:
I don't follow. Is there a reason why our own implementations of Adafactor and SparseAdam are incompatible with `torch.optim.lr_scheduler`? I see that Adafactor already subclasses `torch.optim.Optimizer`. I don't see why it wouldn't be compatible with `lr_scheduler` as long as it uses the same interface as the optimizers in `torch.optim`, which also subclass `torch.optim.Optimizer`.

Reply:
I did consider the LR scheduling from PyTorch but did not found them to add much value over a simple and flexible function based approach. And they are using the epoch terminology which would be confusing in our case.

Regarding Adafactor, the issue I think is that the LR scheduling is defined in the optimizer. It should not be, and it is now very easy to add an additional decay method. I can add that as part of this PR.

Regarding the naming discussed above, I propose:

* `make_optimizer` -> `build_torch_optimizer`
* `Optimizer.from_options` -> `Optimizer.from_opt`